Updated Schema, added route + controller function for finding ID + tests.

Heyjp opened the pull request:
Need some guidance here. 

Tried out using Jest but couldn't get environment variables to work so hard coded the local IP. Is there any way to do this? Also with the Schema, why are we wrapping _id in an object as oppose to letting Mongoose handle it. Or even by just writing . 

```
{
  _id: ObjectId
}
```

zsoltime commented:
I can review your PR when I'm back at home today.

The schema I posted on #40 has an `_id` field which is an ObjectID generated by MongoDB. Although never used it, I think Mongoose uses the same ObjectID when it inserts a document. 

The seed files I uploaded are seed files only and we can use with `mongoimport`. The `_id` `$oid` fields are becoming `"_id" : ObjectId("xxxxxxx")` by importing these JSON files. It might be confusing if you've never used `mongoexport` to export data.

zsoltime commented on server/controller/index.js:
```diff
@@ -25,3 +25,16 @@ exports.getQuestions = async (ctx) => {
 };
+
+exports.getId = async (ctx) => {
+  console.log(ctx.params.id, "this is params")
```
You should use single quotes.

zsoltime commented on server/controller/index.js:
```diff
@@ -25,3 +25,16 @@ exports.getQuestions = async (ctx) => {
+      return ctx.body;
+    })
+    .catch((err) => {
+      ctx.body = "id not recognized";
```
You should use single quotes.

zsoltime commented on server/tests/testSchema.js:
```diff
@@ -0,0 +1,17 @@
+const mongoose = require("mongoose");
```
You should use single quotes.

zsoltime commented on server/models/question.js:
```diff
@@ -1,11 +1,35 @@
 const mongoose = require('mongoose');
 
 const questionSchema = mongoose.Schema({
+  _id: {
```
Not sure if we need the `_id` field here as I mentioned in my [previous post](https://github.com/chingu-voyage3/bears-17/pull/43#issuecomment-351133803).

zsoltime commented on server/models/question.js:
```diff
@@ -1,11 +1,35 @@
+  _id: {
+    $oid: ObjectId,
+  },
+  submitted_at: Date.now(),
```
In your previous PR you set `timestamps: true`. Not sure which one is the preferred way, but we can use `timestamps: { createdAt: 'submitted_at', updatedAt: 'updated_at' }` to change the default names.

zsoltime commented on server/models/question.js:
```diff
@@ -1,11 +1,35 @@
+    name: String,
+    avatar: String,
+  }, 
+  answers: [{
```
We can remove answers now as it has its own collection (PR #42)

zsoltime commented:
Just noticed that I forgot to answer your very first question. To get the content of the `.env` file, you need to add `require('dotenv').config();` on the first line.

However, we should to create a new database first for testing purposes, so tests don't wipe the db we use for development at the moment.

Heyjp commented:
We can create a new DB that is an option, or we can just simply change the collection that we use to store and remove all the data. 

```
const questionSchema = mongoose.Schema(
  {
    title: String,
    question: String,
    user: { default: 'user', type: String },
  },
  {
    timestamps: true,
    collection: 'test',
  },
);
```

Currently the tests run off a local instance of mongodb, on the 'test' collection as well so I don't think it should affect the production DB at least. Not sure if this is the best option happy to be corrected though.

zsoltime commented on server/models/question.js:
```diff
@@ -1,12 +1,23 @@
+    votes: Number,
+    author: {
+      _id: {
+        $oid: ObjectId(),
```
`ObjectId` is not defined and maybe we should add a new `author` schema like in the [docs](http://mongoosejs.com/docs/subdocs.html)?

Heyjp replied:
I agree with this. Would make the DB easier to manage.

Heyjp replied:
I think we can add in a new schema for the `author` but we still need to keep a reference in the `question` schema. Would a string be better or should we just use an `ObjectId` that we would get from the new schema?

zsoltime replied:
The author schema should be no more than a simple `{ name: String, avatar: String }` we don't need to add the `_id` field here, at least I don't think so. But I don't know mongoose :( Could you test it this way?

zsoltime commented on server/models/question.js:
```diff
@@ -1,12 +1,23 @@
-});
+const questionSchema = mongoose.Schema(
+  {
+    submitted_at: Date.now(),
```
Do we still need this one? You've added timestamps already.

Heyjp commented:
So made a slight alteration to the schemas so that we can collect information across both of them if we so choose. This could be useful for say inspecting a profile and wanting to see maybe a list of questions / answers that the user has made. 

Followed the same format as the json files uploaded. Hope my formatting is on point this time.

zsoltime commented:
So is your `authorSchema` the same as `userSchema` would be? I really need to dive into reading more about mongoose population and subdocs, but if I'm correct subdocs are embedded documents (meaning there are no extra queries to get the user's name and avatar) while populated queries are running extra queries behind the scenes.

So if we list 50 questions on a page and we want to display the question title, the date, the author name and avatar, we need 1 query to get the question and 50 more queries to get the author name and avatar. 

We can avoid it by embedding the author name and avatar inside the question and use one single query instead of the 51 above.

Heyjp commented:
Changed the question to have an embedded author like you said. But also removed the questions on the `userSchema` / `authorSchema`. I think that could still benefit from populate on that schema, otherwise you are just making another query to the DB to retrieve all the questions a user might have. Could expand to have Answers as well.

zsoltime commented:
I think you're right, but let's just do that in one of the next sprints and leave it as is. It works as expected and all tests pass. There are a few styling issues I'd like you to fix before I can accept and merge this. I'll comment these in a few minutes.

Anyway, really great job, Jamie, thank you. 👍 😃

zsoltime commented on server/models/question.js:
```diff
@@ -1,12 +1,19 @@
+    },
+  },
+  {
+    timestamps: { createdAt: 'submitted_at', updatedAt: 'updated_at' }
```
There is a missing comma

zsoltime commented on server/models/user.js:
```diff
@@ -0,0 +1,14 @@
+  email: String,
+  avatar: String,
+  country: String,
+  member_since: {type: Date, default: Date.now },
```
There is a missing space before `type`

zsoltime commented on server/tests/questions.test.js:
```diff
@@ -0,0 +1,96 @@
+    done();
+  });
+  
+  test('Should remove a single entry and return an empty array ', async (done) => {
```
There is an extra space after `array`

zsoltime commented on server/tests/questions.test.js:
```diff
@@ -0,0 +1,96 @@
+    done();
+  });
+
+  test('should find and return an entry based on id', async (done) => {
```
`should` should be capitalised (or all the other ones lowercase 😃) to stay consistent.

zsoltime commented on server/tests/testSchema.js:
```diff
@@ -0,0 +1,40 @@
+  User: userModel,
+  Question: questionModel,
+};
+
```
There are two new lines here, should be one only.
